pylint/checkers/typecheck.py

```diff
@@ -2015,6 +2016,20 @@ def visit_compare(self, node: nodes.Compare) -> None:
         if op in {"in", "not in"}:
             self._check_membership_test(right)
 
+    @only_required_for_messages("unhashable-dict-key")
+    def visit_dict(self, node: nodes.Dict) -> None:
+        for k, _ in node.items:
+            if not is_hashable(k):
+                self.add_message("unhashable-dict-key", node=k, confidence=INFERENCE)
+
+    @only_required_for_messages("unhashable-dict-key")
+    def visit_set(self, node: nodes.Set) -> None:
+        for element in node.elts:
+            if not is_hashable(element):
+                self.add_message(
+                    "unhashable-dict-key", node=element, confidence=INFERENCE
```

Comment:
Do you think it could it deserve its own message; for example `unhashable-set-member`?

Reply:
Or maybe we could rename to `unhashable-member` in a backwards-compatible way?

Reply:
That also sounds good.